mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -2319,3 +2374,167 @@ mlir::rock::removeUpperDims(OpBuilder &b, ArrayAttr transformAttrs,
       convertDimNamesToIndices(transformAttrs, removeDimNamesSet);
   return removeUpperDims(b, transformAttrs, removeIndicesSet);
 }
+
+FailureOr<llvm::SmallDenseMap<int64_t, SmallVector<SubDimInfo>>>
+mlir::rock::getLowerSubDimensions(OpBuilder &b, ArrayAttr transformAttrs,
+                                  int64_t dim) {
+  return getLowerSubDimensions(b, transformAttrs, ArrayRef<int64_t>{dim});
+}
+
+FailureOr<llvm::SmallDenseMap<int64_t, SmallVector<SubDimInfo>>>
+mlir::rock::getLowerSubDimensions(OpBuilder &b, ArrayAttr transformAttrs,
+                                  ArrayRef<int64_t> dims) {
+  llvm::SmallDenseMap<int64_t, SmallVector<SubDimInfo>> subDimInfo;
+  TransformMapAttr topMap = cast<TransformMapAttr>(transformAttrs[0]);
+  for (int64_t dim : dims) {
+    // No point of tracing size 1 dimensions
+    if (topMap.getUpperBounds()[dim] == 1) {
+      continue;
+    }
+    LLVM_DEBUG(llvm::dbgs()
+               << "creating subDim of <size=" << topMap.getUpperBounds()[dim]
+               << ",stride=" << 1 << "> @ " << dim << "\n");
+    subDimInfo[dim].push_back({topMap.getUpperBounds()[dim], 1});
+  }
+
+  for (Attribute attr : transformAttrs) {
+    TransformMapAttr trMap = cast<TransformMapAttr>(attr);
+    LLVM_DEBUG(llvm::dbgs() << "analyzing trMap:" << trMap << "\n");
+    // local function to update the next subdim info
+    auto getNextSubDimInfo =
+        [&trMap](const llvm::SmallDenseMap<int64_t, SmallVector<SubDimInfo>>
+                     &currSubDimInfo)
+        -> FailureOr<llvm::SmallDenseMap<int64_t, SmallVector<SubDimInfo>>> {
+      llvm::SmallDenseMap<int64_t, SmallVector<SubDimInfo>> nextSubDimInfo;
+      for (TransformAttr trAttr : trMap.getOps()) {
+        switch (trAttr.getType()) {
+        case TransformType::PassThrough: {
+          llvm::SmallDenseMap<int64_t, int64_t> upperToLower;
+          for (auto [idx, upperDim] : llvm::enumerate(trAttr.getUpperDims())) {
+            const auto lowerDim = trAttr.getLowerDims()[idx];
+            LLVM_DEBUG(llvm::dbgs() << "pt:upper=" << upperDim
+                                    << ",lower=" << lowerDim << "\n");
+            upperToLower[upperDim] = lowerDim;
+          };
+          for (auto [dim, subDimInfo] : currSubDimInfo) {
+            if (upperToLower.contains(dim)) {
+              LLVM_DEBUG(llvm::dbgs() << "remapping:" << dim << " to "
+                                      << upperToLower[dim] << "\n");
+              nextSubDimInfo[upperToLower[dim]] = subDimInfo;
+            }
+          }
+        } break;
+        case TransformType::Merge: {
+          SmallVector<int64_t> subDimStrides = getStrides(trAttr.getParams());
+          for (size_t subDim = 0; subDim < trAttr.getParams().size();
```

Comment:
This could be a `llvm::zip()` or an `llvm::enumerate()` if you do need the index

Reply:
thanks! it did